File: 1335. Minimum Difficulty of a Job Schedule/1335. Minimum Difficulty of a Job Schedule top-down 1.java
import java.util.Arrays;

class Solution {
    private int n, d;
    private int[][] memo;
    private int[] jobDifficulty, hardestJobRemaining;

    private int dp(int i, int day) {
        // base case
        if (day == d)
            return hardestJobRemaining[i];

        if (memo[i][day] == -1) {
            int hardest = 0;
            int best = Integer.MAX_VALUE;
            // recurrence relation
            for (int j = i; j < n - (d - day); j++) {
                hardest = Math.max(hardest, jobDifficulty[j]);
                best = Math.min(best, hardest + dp(j + 1, day + 1));
            }
            memo[i][day] = best;
        }

        return memo[i][day];

    }

    public int minDifficulty(int[] jobDifficulty, int d) {
        this.n = jobDifficulty.length;
        this.d = d;
        if (n < d)
            return -1;

        this.jobDifficulty = jobDifficulty;
        memo = new int[n][d + 1];
        hardestJobRemaining = new int[n];

        int hardest = 0;

        for (int i = n - 1; i >= 0; i--) {
            // set hardestJobRemaining
            hardest = Math.max(hardest, jobDifficulty[i]);
            hardestJobRemaining[i] = hardest;

            // fill memo with -1
            Arrays.fill(memo[i], -1);
        }

        return dp(0, 1);
    }
}

File: 1335. Minimum Difficulty of a Job Schedule/1335. Minimum Difficulty of a Job Schedule top-down 1.py
from functools import lru_cache
from typing import List


class Solution:
    def minDifficulty(self, jobDifficulty: List[int], d: int) -> int:
        n = len(jobDifficulty)
        if n < d:
            return -1

        hardest_job_remaining = [0]*n
        hardest_job = 0

        for i in range(n-1, -1, -1):
            hardest_job = max(hardest_job, jobDifficulty[i])
            hardest_job_remaining[i] = hardest_job

        @lru_cache(None)
        def dp(i: int, day: int) -> int:
            # base case
            if d == day:
                return hardest_job_remaining[i]

            # recurrence relation
            best = float('inf')
            hardest = 0

            for j in range(i, n-(d-day)):
                hardest = max(hardest, jobDifficulty[j])
                best = min(best, hardest + dp(j+1, day+1))
            return best
        return dp(0, 1)


# jobDifficulty = [6, 5, 4, 3, 2, 1]
# d = 2
jobDifficulty = [9, 9, 9]
d = 4
# jobDifficulty = [1, 1, 1]
# d = 3
s = Solution()
print(s.minDifficulty(jobDifficulty, d))
